### DISCUSSION TITLE
Connection without user input

### DISCUSSION THREAD

**Discussion by hillsh** [capacitor-community/bluetooth-le#605] [Q&A]
I'm writing an app that requires connecting to a BLE device.  I want to store the device information (BleDevice) in storage after I make my first successful connection, and thereafter on starting the app just connect using the information in storage.  However I am getting connection errors indicating that the device is not found. However if I run requestDevice first then all is OK.  This requires user input to choose a device from the scanning dialog.  Is there a way to circumvent this?

**Comment by mutasimissa** [comment]
Hello
What are you storing exactly to reconnect to the device? Do you have some code to look at?

**Reply by hillsh** [reply]
Here is the code I am using:
  BleClient.initialize()
      .then(() => {
        console.log('Initialized OK');
        this.device = this.obdDevice$.value;
        console.log('Name: ' + this.device.name);
        if (this.device.name === undefined) {
          BleClient.requestDevice({
            //          name: 'OBDII',
            services: [],
          })
            .then((dev) => {
              console.log('Got a device');
              this.device = dev;
              this.obdDevice$.next(dev);
              this.sharedObdDevice$.pipe(take(1))
              .subscribe(val=>{console.log("New name: "+ val.name)});
              BleClient.connect(this.device.deviceId)
                .then(() => {
                  this.readyToGo$.next(true);
                  console.log('All good!');
                  this.showToast('Connected!');
                })
                .catch((error) => {
                  this.showToast(
                    'Cannot connect! Close the app. Error = ' + error
                  );
                  return;
                });
            })
            .catch((error) => {
              this.showToast('No OBD Device! Close the app. Error: ' + error);
              return;
            });
        } else {
          console.log("Connecting to "+ this.device.deviceId);
          BleClient.connect(this.device.deviceId)
            .then(() => {
              this.readyToGo$.next(true);
              console.log('All good!');
              this.showToast('Connected!');
            })
            .catch((error) => {
              this.showToast('Cannot connect! Close the app. Error = ' + error);
              return;
            });
        }
      })
      .catch(() => {
        this.showToast(
          'No Bluetooth! Close the app, enable Bluetooth and restart.'
        );

**Reply by hillsh** [reply]
I have noticed that when using requestDevice or requestLEScan it often takes quite a while (several seconds) for the device to show up in the scanning dialog. So maybe when I try to connect to the device with the stored deviceId, the BLE system does not know where the device is?  I don't know how these things work "under the hood".

**Comment by mutasimissa** [comment]
I think the problem is related to fetching the deviceId from storage, did you make sure you are doing that correctly and that you can console.log the deviceId?
Another thing is that it's much better to split your code into different functions with separate concerns and to use async/await with try/catch blocks instead of chaining then and catch.
see this example. it's messy also but I think you will find what you need.
https://github.com/sourcya/ble-tester/blob/main/src/pages/bleTest.js

**Reply by hillsh** [reply]
I should point out that I am testing this on a Chrome browser on a Win 11 desktop.  I have noticed that the device ID for the device I am connecting with for testing is not constant.

**Reply by hillsh** [reply]
Hello, I hope someone is following this question.  I have noticed that the device I am connecting to (on the web, Chrome browser, Win 11 desktop) which is a Samsung M7 monitor, does not maintain a constant device ID, which means that saving the device ID in memory and attempting to reconnect would never work.  Is this the expected behaviour for BLE devices on the web?  I know that in Android the device ID is the MAC, which I assume would not be changing.

**Reply by hillsh** [reply]
Reply to Mutasimissa: yes, I can console log the device ID, but as I pointed out in my earlier comment, there seems to be no point in saving the device ID in memory, as it is not constant.  I am studying your code - I am a little leery of async/await, because my use of it in the past has not worked out that well.  For example, consider the following fragment:
      await BleClient.initialize();
      await BleClient.disconnect(this.state.selectedDevice.deviceId);
      await BleClient.connect(this.state.selectedDevice.deviceId, (id) =>
        console.log(`Device ${id} disconnected!`)
Why would the system not start initialize, then immediately move on to start disconnect, then immediately move on the start connect, since none of these statements depends on the result of the previous ones?

**Reply by hillsh** [reply]
Continuing with this thread: Mutasimissa, i have been looking at your code and I see that whenever you do anything at all the BLE device (e.g. read, write, set up notifications, etc), you initialize BleClient, disconnect, re-connect, and then carry out whatever operation you want to do.  Is this accepted best practice for working with BLE devices?
On the question of saving device ID and using it at a later time to connect, I see from this resource:
https://webbluetoothcg.github.io/web-bluetooth/#bluetooth-device-identifiers
that the whole concept of a deviceId is extremely complex.  It looks like whenever you want to work in an app with a BLE device you have to call requestDevice or getDevices first.  Also if you call requestDevice you have to know the service UUIDs you are going to use beforehand and specify them in optionalServices  or you will not be allowed to connect.  Does this sound correct to you?

**Reply by mutasimissa** [reply]
I doubt that the deviceId changes on the same piece of hardware, I think you need to do more testing.
Regarding my code and initializing -> disconnect -> connect on each operation, it's not the best practice but my goal was to build a tester to test mainly ble printers, some devices have problems after connecting and I couldn't connect any device before I disconnect, so what I did was an ugly shortcut, I encourage you to do a better lifecycle.

**Reply by peitschie** [reply]
> I doubt that the deviceId changes on the same piece of hardware

Worth noting that there is a privacy configuration for BLE peripherals which does exactly this, and rotates the device ID every 15mins or so.

Check out private addresses in https://novelbits.io/bluetooth-address-privacy-ble/

**Comment by mutasimissa** [comment]
@peitsche thanks for the information never knew that! Also the novelbits.io itself seems a very good option to learn bluetooth development never saw specialized courses for ble development!
